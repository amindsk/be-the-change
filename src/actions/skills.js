import axios from 'axios';
const instance = axios.create({
    baseURL: 'http://localhost:3000/'
});

//Redux Action: Add_SKILL
//To add a new skill
export const addSkill = (skill) => ({
    type: 'ADD_SKILL',
    skill
});

export const startAddSkill = (skillData = {}) => {
    return (dispatch) => {
        const {
            name = '',
            experience = ''
        } = skillData;
        const skill = { name, experience };
        return instance.post('/skills', skill ).then(({ data }) => {
            dispatch(addSkill(data));
        });

    };
};

//Redux Action: REMOVE_SKILL
//To remove a skill by id
export const removeSkill = ( { id } = {}) => ({
    type: 'REMOVE_SKILL',
    id
});

export const startRemoveSkill = ( { id } = { } ) => {
    return (dispatch) => {
        return instance.delete(`/skills/${id}`).then(() => {
            dispatch(removeSkill( { id } ));
        });
    };
};

export const setSkills = (skills) => ({
    type: 'SET_SKILLS',
    skills
});

export const startSetSkills = () => {
    return (dispatch) => {
        return instance.get('/skills').then((data) => {
            const skills = data.data;
            dispatch(setSkills(skills));
        });
    };
};