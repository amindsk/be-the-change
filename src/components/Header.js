//To create an app component
import React from 'react';

//Header component
const Header = () => (
    //JSX of the component
    <header>
        <h1>Skill Editor</h1>
    </header>
);

//Exporting component
export default Header;