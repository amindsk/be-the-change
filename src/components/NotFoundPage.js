//To create an app component
import React from 'react';
//To create a link back to Home page
import { NavLink } from 'react-router-dom';

//Page not found component for invalid URLs
const NotFoundPage = () => (
    //Component JSX
    <div>
        404 Not Found<NavLink activeClassName="home" to="/">Home</NavLink>
    </div>
);

//Exporting the component
export default NotFoundPage;