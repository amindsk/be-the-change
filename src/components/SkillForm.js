//Importing React to make component
import React from 'react';
//Class based component SkillForm to add skill
export default class SkillForm extends React.Component {
    //Local states of the component
    state = {
        name: '',
        experience: '',
        //To show error on wrong inputs/values provided
        error: ''
    };
    //Binding input value of skill name with component's local state 
    onNameChange = (e) => {
        //Get the value of the input to prevent using presist()
        const name = e.target.value;
        //Bind value with slocal state
        this.setState(() => ({ name }));
    }
    //Binding select value of skill experience with component's local state
    onExperienceChange = (e) => {
        //Get the value of the select to prevent using presist()
        const experience = e.target.value;
        //Bind value with slocal state
        this.setState(() => ({ experience }));
    }
    //Handle for submit event
    onSubmit = (e) => {
        //Prevent default behaviour of the form on submit ( Prevent page from refreshing )
        e.preventDefault();
        //If skil name or skill experience are not set (empty)
        if(!this.state.name || !this.state.experience) {
            //Display error to provide values off the skill name and experience
            this.setState(() => ({ error: 'Please provide skill name and experience' }));
            //If name of the skill is less than 4 or greater that 255 characters
        } else if(this.state.name.length < 4 || this.state.name.length > 255) {
            //Display error to provide name of the skill within given range
            this.setState(() => ({ error: 'Please provide a skill name of minimum 4 characters' }));
            //If values of name and experience are correct
        } else {
            //Erase error if any
            this.setState(() => ({ error: '' }));
            //Call the on submit prop to send data back to parent component
            this.props.onSubmit({
                //Set skill object
                name: this.state.name,
                experience: this.state.experience
            });
            //Erase previously entered values
            this.setState({ name: '', experience: '', error: '' });
        }

    }
    //Render the JSX
    render() {
        //Return JSX
        return (
            <div>
                {/* Show error if any */}
                {this.state.error && <p>{this.state.error}</p>}
                {/* Add skill form */}
                <form onSubmit={this.onSubmit}>
                {/* Skill name */}
                    <input
                    type="text"
                    placeholder="Skill Here"
                    autoFocus
                    value={this.state.name}
                    onChange={this.onNameChange}
                    />
                    {/* Skill experience */}
                    <select
                    value={this.state.experience}
                    onChange={this.onExperienceChange}
                    >
                        <option value="" disabled selected>EXPERIENCE</option>
                        <option value="< 1 year">&lt; 1 year</option>
                        <option value="1 - 3 years">1 - 3 years</option>
                        <option value="3 - 5 years">3 - 5 years</option>
                        <option value="5 - 7 years">5 - 7 years</option>
                        <option value="7+ years">7+ years</option>
                    </select>
                    {/* Add skill button */}
                    <button>ADD SKILL</button>
                </form>
            </div>
        );
    }
}