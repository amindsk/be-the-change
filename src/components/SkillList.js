//To create a new component
import React from 'react';
//Importing connect to dispatch actions and read from store
import { connect } from 'react-redux';
//Importing a SkillListItem component
import SkillListItem from './SkillListItem';
//Importing the select to fetch skills ordered by ids
import selectSkills from '../selectors/skills';
//Component
export const SkillList = (props) => (
    <div className="items-list">
        {
            //Display No Skilssif there is not any skill
            props.skills.length === 0 ? (
            <p>No Skills</p>
            ) : (
                //Render skills if any
                props.skills.map((skill, index) => {
                    //To apply css class on first five skills
                    return index < 5 ?
                     <SkillListItem key={skill.id} firstFive={true} {...skill} />
                    : <SkillListItem key={skill.id} {...skill} />;
                })
            )
        }
    </div>
);

//Mapping redux store states to this component's props
const mapStateToProps = (state) => {
    return {
        skills: selectSkills(state.skills, state.filters)
    };
};

//Exporting component
export default connect(mapStateToProps)(SkillList);