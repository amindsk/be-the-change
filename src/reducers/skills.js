//Redux Reducer: Skills
//Initial/Default values for skills state
const skillsReducerDefaultState = [];
//Embedding actions to skill's reducer
export default (state = skillsReducerDefaultState, action) => {
    switch(action.type) {
        case 'ADD_SKILL':
            return [
                ...state,
                action.skill
            ];
        case 'SET_SKILLS':
            return action.skills
        case 'REMOVE_SKILL':
            return state.filter(( { id } ) => id !== action.id);
        default:
        return state
    }
};