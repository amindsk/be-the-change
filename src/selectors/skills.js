//Redux Selector: To get a sorted skills based on ids
export default (skills, { sortBy }) => {
    return skills ? skills.sort((a, b) => {
        if (sortBy === 'id') {
            return a.id > b.id ? 1 : -1;
        }
    }) : skills;
};