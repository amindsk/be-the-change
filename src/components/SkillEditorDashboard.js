//To create an app component
import React from 'react';
//Importing connect to dispatch actions and read from store
import { connect } from 'react-redux';
//Importing SkillList component
import SkillsList from './SkillsList';
//Importing SkillForm component
import SkillForm from '../components/SkillForm';
//Importing addSkill action from redux
import { addSkill } from '../actions/skills';
//Base component for the app
const SkillEditorDashboard = (props) => (
    //Component JSX
    <div>
        {/* Embedding SKillForm component to add new skill */}
        <SkillForm onSubmit={(skill) => {
            // Passing skill to add to the store on submit of form
            props.dispatch(addSkill(skill));
        }} />
        {/* Embedding SkillList component to render all of th elsit items */}
        <SkillsList />
    </div>
);

//Exporting the component with connect to dispatch actions and read from store
export default connect()(SkillEditorDashboard);