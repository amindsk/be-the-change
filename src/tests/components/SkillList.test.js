import React from 'react';
import { shallow } from 'enzyme';
import { SkillList } from '../../components/SkillList';
import skills from '../fixtures/skills';

test('Should render SkillList correctly', () => {
    const wrapper = shallow(<SkillList skills={skills} />);
    expect(wrapper).toMatchSnapshot();
});

test('Should render SkillList with empty message', () => {
    const wrapper = shallow(<SkillList skills={[]} />);
    expect(wrapper).toMatchSnapshot();
});