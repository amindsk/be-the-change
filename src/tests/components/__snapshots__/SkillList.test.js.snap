// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render SkillList correctly 1`] = `
<div>
  <Connect(SkillListItem)
    experience="7+ years"
    id={3}
    key="3"
    name="ReactJS"
  />
  <Connect(SkillListItem)
    experience="3 - 5 years"
    id={1}
    key="1"
    name="NodeJS"
  />
  <Connect(SkillListItem)
    experience="1 - 3 years"
    id={2}
    key="2"
    name="PostgreSQL"
  />
</div>
`;

exports[`Should render SkillList with empty message 1`] = `
<div>
  <p>
    No Skills
  </p>
</div>
`;
